elephant/signal_processing.py

```diff
@@ -239,6 +239,82 @@ def _design_butterworth_filter(Fs, hpfreq=None, lpfreq=None, order=4):
         return filtered_data
 
 
+def wavelet_transform(signal, freq, nco, fs):
+    """
+    Compute the wavelet transform of a given signal with Morlet mother wavelet.
+    The definition of the wavelet is based on Le van Quyen et al. J
+    Neurosci Meth 111:83-98 (2001).
+
+    Parameters
+    ----------
+    signal : 1D array_like
+        Signal to be wavelet-transformed
+    freq : float
+        Center frequency of the Morlet wavelet.
+    nco : float
+        Size of the mother wavelet (approximate number of cycles within a
+        wavelet). A larger nco value leads to a higher frequency resolution but
+        a lower temporal resolution, and vice versa. Typically used values are
+        in a range of 3 - 8.
+    fs : float
+        Sampling rate of the signal.
+
+    Returns
+    -------
+    signal_trans: 1D complex array_like
+        Wavelet-transformed signal
+    """
+    # Morlet wavelet generator (c.f. Le van Quyen et al. J Neurosci Meth
+    # 111:83-98 (2001))
+    def _morlet_wavelet(freq, nco, Fs, N):
+        sigma = nco / (6. * freq)
+        t = (np.arange(N, dtype='float') - N / 2) / Fs
+        if N % 2 == 0:
+            t = np.roll(t, int(N / 2))
+        else:
+            t = np.roll(t, int(N / 2) + 1)
+        return np.sqrt(freq) * np.exp(
+            -(t * t) / (2 * sigma ** 2) + 1j * 2 * np.pi * freq * t)
+
+    data = np.asarray(signal)
+    # When the input is AnalogSignal, the axis for time index (i.e. the
+    # first axis) needs to be rolled to the last
+    if isinstance(signal, neo.AnalogSignal):
+        data = np.rollaxis(data, 0, len(data.shape))[0]
+
+    # check whether the given central frequency is less than the Nyquist
+    # frequency of the signal
+    if freq >= fs / 2:
+        raise ValueError(
+            "freq must be less than the half of Fs "+
+            "(sampling rate of the original signal)")
+
+    # check if nco is positive
+    if nco <= 0:
+        raise ValueError("nco must be positive")
+
+    N = len(data)
+    # the least power of 2 greater than N
+    N_pow2 = 2 ** (int(np.log2(N)) + 1)
+
+    # zero-padding to a power of 2 for efficient convolution
+    tmpdata = np.zeros(N_pow2)
+    tmpdata[0:N] = data
+
```

Comment:
I think that there should be an option to turn off zero-padding. The problem with zero-padding is that it changes the total energy of the spectral estimate and there are cases in which the energy is an important quantity.

Reply:
Good point.
I implemented zero_padding option to turn on/off zero-padding.